python/tests/server/test_worker.py

```diff
@@ -669,7 +719,20 @@ def await_predict(self, state: PredictState):
         state=consumes(predict_pending),
     )
     def cancel(self, state: PredictState):
-        self.worker.cancel()
+        self.worker.cancel(tag=state.tag)
+
+        if state.canceled:
+            # if this has previously been canceled, we expect no Cancel event
+            # sent to the child
+            assert not self.child_events.poll(timeout=0.1)
+        else:
+            assert self.child_events.poll(timeout=0.5)
+            e = self.child_events.recv()
+            assert isinstance(e, Envelope)
+            assert isinstance(e.event, Cancel)
+            assert e.tag == state.tag
+            assert self.child.cancel_sent
```

Comment:
I'm not sure how I feel about asserting on what we sent to the child -- it feels like it might make things a bit more brittle than before. Was there something specific that prompted adding this?

Reply:
we don't have to assert, but listening to the child events is how we ensure we synchronize the test with the worker and wait for the worker event consumer thread to do its thing.

This doesn't matter too much here, but it does matter in predict() where we need to wait until the PredictionInput event has been sent to the child - if we don't do that, we can send a Done event for a prediction before the PredictionInput event is sent, which breaks Worker assumptions.

And once we're monitoring the child worker for the PredictionInput event we need to monitor the Cancel event here too because we have to monitor the whole connection.

Reply:
I'll add a comment to make this clearer in predict()